lib/pbench/test/unit/server/test_unpack_tarballs.py

```diff
@@ -0,0 +1,293 @@
+import os
+from pathlib import Path
+
+import pytest
+
+from pbench.common.logger import get_pbench_logger
+from pbench.server.filetree import FileTree, TarballUnpackError
+from pbench.server.unpack_tarballs import UnpackTarballs
+
+
+@pytest.fixture()
+def make_logger(server_config):
+    """
+    Construct a Pbench Logger object
+    """
+    return get_pbench_logger("TEST", server_config)
+
+
+class TestUnpackTarballs:
+    tarlist = ["/mock/ABC/TO-UNPACK/A.tar.xz", "/mock/ABC/TO-UNPACK/B.tar.xz"]
+    min_size, max_size, tar_size = 0, 1000, 586
+
+    def mock_stat(self, *args, follow_symlinks=False):
+        """Mocked replacement for Path.stat()"""
+        StatResult = namedtuple("StatResult", ["st_mode", "st_size"])
+        return StatResult(st_mode=16893, st_size=self.tar_size)
+
+    @pytest.mark.parametrize(
+        "min_size, max_size, total", [(586, 1000, 2), (0, 1000, 2), (0, 586, 0)]
+    )
+    def test_tarball_size(
+        self, monkeypatch, server_config, make_logger, caplog, min_size, max_size, total
+    ):
+        """Test for the valid size and path of the tarball"""
+
+        def mock_resolve(self: Path, strict: bool = False):
+            """
+            Mock replacement for Path.resolve()
+
+            Args:
+                strict: Boolean value to make sure path exist if True.
+
+            Returns:
+                Raising FileNotFoundError Exception
+            """
+            assert strict
+            raise FileNotFoundError(f"No such file or directory: '{self}'")
+
+        monkeypatch.setattr(Path, "glob", lambda path, args: self.tarlist)
+        monkeypatch.setattr(Path, "resolve", mock_resolve)
+        monkeypatch.setattr(Path, "stat", self.mock_stat)
+        obj = UnpackTarballs(server_config, make_logger)
+        result = obj.unpack_tarballs(min_size, max_size)
+        if result.total > 0:
+            for tar in self.tarlist:
+                assert (
+                    f"Tarball link, '{tar}', does not resolve to a real location: No such file or directory: '{tar}'"
+                    in caplog.text
+                )
+        assert result.total == total
+        assert result.success == 0
+
+    def test_unpack_md5_error(self, monkeypatch, server_config, make_logger, caplog):
+        """Show that when md5 file path cannot be accessed it throws a
+        FileNotFoundError Exception"""
+
+        def mock_md5sum(filepath: Path):
+            """
+            Mock of md5sum function
+
+            Args:
+                filepath: Tarball file path
+
+            Returns:
+                Raises FileNotFoundError Exception
+            """
+            raise FileNotFoundError(f"No such file or directory: '{filepath}.md5'")
+
+        monkeypatch.setattr("pbench.server.utils.md5sum", mock_md5sum)
+        file_tree = FileTree(server_config, make_logger)
+        obj = UnpackTarballs(server_config, make_logger)
+        for tar in self.tarlist:
+            with pytest.raises(Exception):
```

Comment:
at line 114, `FileTree.unpack` is being mocked and the exception that is raised there is already handled in `unpack module`, we are re-raising and handling error again.

Reply:
In the case of this test here (`test_unpack_md5_error()`), when the mock for `md5sum()` raises `FileNotFoundError`, I presume that that propagates out through `get_tarball_md5()` to `unpack()` where it is caught and reraised.  Therefore, the test should receive a `FileNotFoundError`, and so we should expect that here instead of `Exception`.

In the case of `TarballUnpackError()`, when the mock for `FileTree.unpack()` raises `TarballUnpackError`, the exception is caught and reraised...but, what is reraised should be a `TarballUnpackError`, even though it was caught using an `except Exception` clause:  the type specified in the `except` statement is for matching purposes -- it doesn't change the instance which is actually caught.  And, the bare `raise` should reraise the _original instance_ of the exception.  Therefore, the test should receive a `TarballUnpackError`, and so we should expect that there instead of `Exception`.